Firestore/core/src/core/firestore_client.cc

```diff
@@ -263,7 +275,13 @@ void FirestoreClient::Dispose() {
 
 void FirestoreClient::TerminateAsync(StatusCallback callback) {
   worker_queue_->EnterRestrictedMode();
-  worker_queue_->EnqueueEvenWhileRestricted([this, callback] {
+  worker_queue_->EnqueueEvenWhileRestricted([&, this, callback] {
+    // Make sure this `FirestoreClient` is not destroyed during
+    // `TerminateInternal`, so that `user_executor_` stays valid.
+    auto self = shared_from_this();
```

Comment:
This is arguably unnecessary -- `api::Firestore`, locked below, would contain a pointer to the client. I'm not sure there's a case where `weak_firestore_` would return null here (perhaps in unit testing), but it seems like a safer, if perhaps redundant, approach.